### ISSUE TITLE
ability to pass arguments to other applications

### ISSUE THREAD

**Issue by karinaardos** [ssborbis/ContextSearch-web-ext#459]
The ability to pass arguments to other applications is very lacking.
I'm currently using the External Application Button extension

**Comment by ssborbis** [comment]
That would be a cool addition.

I have a python3-based native app ready to go. Basic testing looks good.

I've tested the following so far

"open in firefox"
"open in chrome"
"open in gimp"

![externalProgram](https://user-images.githubusercontent.com/24310997/153655700-c44ff991-8f1e-44ee-8cf4-f2424d9fb947.jpg)

**Comment by karinaardos** [comment]
thank you, you are a very responsive programmer.
and I want to tell you that you have created the best of the best supplement. Your add-on works flawlessly and performs its functions perfectly.
I meant that the add-on can transfer the selected text to other windows 10 applications.

many programs have a command line

**Comment by ssborbis** [comment]
> I meant that the add-on can transfer the selected text to other windows 10 applications.
many programs have a command line

What programs are you looking to open from this addon? I can do some testing locally.

**Comment by ssborbis** [comment]
I am loving this addition BTW. I do a lot of graphic design and opening images straight into GIMP is already saving me time. Thanks for the suggestion!

I'll reopen the old native app page I have and replace the old code with the new code that supports launching external apps over the next week or 2.

**Comment by karinaardos** [comment]
> > Я имел в виду, что дополнение может передавать выделенный текст другим приложениям Windows 10. 
> > многие программы имеют командную строку
> 
> Какие программы вы хотите открыть из этого дополнения? Я могу сделать некоторые тесты на месте.

Instant file search program on all drives (Everything.exe)

I would like to transfer the selected text to the (AkelPad.exe) notepad  - "but it seems to me that there will be difficulties with this program"

**Comment by ssborbis** [comment]
> I would like to transfer the selected text to the (AkelPad.exe) notepad - "but it seems to me that there will be difficulties with this program"

I'm sure there are ways.

If you wanted a new file with the text, you could launch a script that saves the selected text to a temporary file, then launches AkelPad.exe

**Comment by ssborbis** [comment]
I have a beta in the release section ( Firefox ) and the python native app can be found [here](https://github.com/ssborbis/ContextSearch-Native-App) if you want to test.

**Comment by karinaardos** [comment]
Unfortunately I didn't get anything.

I did all the steps according to the instructions but I only see an error message at the top of firefox 
(Native messaging app is not connected. See support site for more details.)

Here's what I did.
i use windows 10
firefox portable version 

1. Created a folder in a directory named (ContextSearch-webext)
 c:\Users\MYNAME\AppData\Roaming\ContextSearch-webext\

2. Copied two files into this folder
 c:\Users\MYNAME\AppData\Roaming\ContextSearch-webext\ContextSearch.py
 c:\Users\MYNAME\AppData\Roaming\ContextSearch-webext\contextsearch_webext.json

3. Opened in notepad (contextsearch_webext.json) and changed the way
 {
"name": "contextsearch_webext",
"description": "Launch applications from ContextSearch-webext",
"path": "c:\\Users\MYNAME\\AppData\\Roaming\\ContextSearch-webext\\ContextSearch.py",
"type": "stdio",
"allowed_extensions": [ "{5dd73bb9-e728-4d1e-990b-c77d8e03670f}" ]
}

4. Added a line to the registry (and it appeared in the registry) 
 REG ADD "HKCU\Software\Mozilla\NativeMessagingHosts\contextsearch_webext" /ve /t REG_SZ /d "c:\Users\MYNAME\AppData\Roaming\ContextSearch-webext\contextsearch_webext.json" /f

5. In (Context Search 1.35.0beta) Created an external application launch button
 Name.
 Everything

 The path is shown like this.
 d:\Total Commander\Utilites\Everything\Everything.exe

I can't figure out what I'm doing wrong

**Comment by ssborbis** [comment]
There's a problem with Windows inconsistently launching python `*.py` files as executables. The workaround is to make  a `*.bat` file instead. I had to do this on my Windows 10 install to get this to work.

In contextsearch_webext.json, change ...
```
"path": "c:\Users\MYNAME\AppData\Roaming\ContextSearch-webext\ContextSearch.py",
```
to
```
"path": "c:\Users\MYNAME\AppData\Roaming\ContextSearch-webext\ContextSearch.bat",
```

In folder `c:\Users\MYNAME\AppData\Roaming\ContextSearch-webext\`, create a new file named `ContextSearch.bat` with the following content

```
@echo off
C:\Users\MYNAME\AppData\Local\Programs\Python\Python309\python.exe ContextSearch.py
```

Change line 2 to point to your python.exe

**Comment by ssborbis** [comment]
Also, I just noticed you're using a portable version of Firefox. I don't know how that will effect things. I will test the portable version on my machine.

**Comment by ssborbis** [comment]
The portable version appears to be reading the registry keys correctly.

At any point, did you receive a prompt to grant the `nativeMessaging` permission during your testing?

Check that the `nativeMessaging` permission is granted under `Addons -> ContextSearch web-ext -> Permissions` tab
![permission](https://user-images.githubusercontent.com/24310997/154122777-d429dfec-7919-407f-a9c2-8b85efc3ce67.jpg)

**Comment by ssborbis** [comment]
I added an install.py script to the repo. This *should* work to install to all platforms. Run it from an administrative prompt.

**Comment by ssborbis** [comment]
Thanks for the video. That's very helpful.

Everything looks like it installed ok. The only thing I couldn't check was the registry keys. Could you run install.py again from the console and paste the output here?

Lets see if the browser is in fact communicating with the python app.

Open the CS options page, then the dev console ( Shfit+Ctrl+K ) and enter this

`browser.runtime.sendNativeMessage("contextsearch_webext", {verify:true}).then(r => console.log(r));`

If you get an error, the native messaging API is not finding the app

If you get `true`, maybe try a simpler command. 

From the same console, try running `browser.runtime.sendNativeMessage("contextsearch_webext", {path:"notepad"});`

**Comment by ssborbis** [comment]
I think python is requiring double-backslashes for windows locations, or a single forward-slash

So, if I were using the full path for, say, notepad

```"C:\\Windows\\system32\\notepad.exe"``` 
or 
```"C:/Windows/system32/notepad.exe"``` 
would work

but not 
```"C:\Windows\system32\notepad.exe"```

**Comment by karinaardos** [comment]
You were right about the "\\" count

I pasted the path from your example:
"C:\\Windows\\system32\\notepad.exe"
and it works

I noticed that there is a space in my path and I moved the program to the root of the disk to get rid of spaces
e:\\Everything\\Everything.exe -s {searchTerms}
and it started up fine with the right arguments

Doesn't work if:
d:\Total Commander\Utilites\Everything\Everything.exe -s "{searchTerms}"
"d:\Total Commander\Utilites\Everything\Everything.exe" -s "{searchTerms}"

Working
"d:\\Total Commander\\Utilites\\Everything\\Everything.exe" -s {searchTerms}

You did it wonderfully. thanks a lot

**Comment by ssborbis** [comment]
The quotes issues seems to be taken care of in newer versions of `ContextSearch.py`.

You can download the package again and run `install.py` or replace `ContextSearch.py` in `%user%\AppData\Roaming\ContextSearch-webext` with the new version

**Comment by karinaardos** [comment]
Yes, it's even better now.

**Comment by karinaardos** [comment]
I can't run files with extension (.bat)

Is it so intended?

**Comment by ssborbis** [comment]
> I can't run files with extension (.bat)

I updated ContextSearch.py with a slightly different call to open programs. It seems to be running .bat files for me.

If you need to see the terminal output, use `start` in front of the command like so:

```
start c:/users/mclovin/desktop/test.bat
```

**Comment by ssborbis** [comment]
You may also run into the issue of your .bat file not working because the current working directory is not the same as the directory the .bat is in. If need be, I can probably add a cwd field, but you should be able to get around it with something like this:

```
start cmd /k "cd c:/users/mclovin/Desktop && test.bat"
```

**Comment by karinaardos** [comment]
You never cease to make me happy. Bat files now run perfectly. thank you so much

**Comment by karinaardos** [comment]
I write and laugh.
because I wanted to offer you this, but you seem to have read my thoughts.
As a result, you created a very cool add-on

**Comment by ssborbis** [comment]
1.35.0 has been released to AMO and Chromestore.

**Comment by logusergithub** [comment]
Remarkable new feature. Thanks. I tested it with PotPlayer, MPV, Yt-dl, Firefox (other profile), Edge, JDownloader and Foobar2000.
I don't know why the extension tell me that 'Messenger app is not connected' despite everything works fine, but I probably miss some understanding.
Unless you already can do, is it possible to add an option inside the Launcher 's settings to directly use the address bar URL when nothing is selected? This mainly concerns a faster use with the Toolbar.

**Comment by ssborbis** [comment]
> I don't know why the extension tell me that 'Messenger app is not connected' despite everything works fine, but I probably miss some understanding

That is odd. If everything is working correctly, it should should the version of the native app where the messaging you're seeing is. I wonder if you have a slightly older version of the native app installed.

![native](https://user-images.githubusercontent.com/24310997/155772993-f4a17b11-e883-49ed-a801-8309d409e33e.jpg)

> Unless you already can do, is it possible to add an option inside the Launcher 's settings to directly use the address bar URL when nothing is selected? This mainly concerns a faster use with the Toolbar.

I can add additional variables besides {searchTerms} to be replaced. 

I know {domain} and {subdomain} are replaced ( for search engines at least, not sure about app launchers ). I could add {url}. That would effectively give you what you want.

**Comment by logusergithub** [comment]
Oh, I see in your picture v2.09, I installed v2.00 from the related Tags.
As for the feature request that would be nice, I've usually always used a "Toolbar (list)" like tool that uses URLs on the address bar and a Context Menu for the rest.  Thanks.

**Comment by ssborbis** [comment]
Code is in the dev branch

Next release, you can simply add this to your Modify Terms regex
```
/^$/{url}/g
```

It will replace empty terms with {url} which is then replaced with the url of the current active tab

**Comment by logusergithub** [comment]
This is perfect.  At this rate in the near future I imagine that if I will use the Toolbar to look for 'coffee' a cup from the screen will be released. Just for the record, at the moment your extension can support over 500 search engines without slowdowns.

**Comment by ssborbis** [comment]
v1.36 is live with the `{url}` param added. Let me know how it works out

**Comment by logusergithub** [comment]
> v1.36 is live with the {url} param added. Let me know how it works out

I tested on 1.36 (new profile) and 1.37b1 (man profile):
"C:\Program Files\Mozilla Firefox\firefox.exe" -p "default" -no-remote "{url}"
It workd well. Thanks.  NB:
- for other users... if you 'Copy' an 'App launcher' is not like a shortcut so you have to modify the copy too
- as I wrote in the other post, seems that 'Modify terms regex' in 'App launchers' are not saved
- despite everything already worked fine, I forgot to tell you that now I see your picture after installed v2.09

**Comment by ssborbis** [comment]
> as I wrote in the other post, seems that 'Modify terms regex' in 'App launchers' are not saved

It is saved, but is not displayed when you next open the edit modal. It's fixed now locally. Thanks.

> for other users... if you 'Copy' an 'App launcher' is not like a shortcut so you have to modify the copy too

This is true. Only OpenSearch engines get the shortcut option. Everything else is an independent copy. ( scripts, app launcher, site search, firefox engines, etc ).

**Comment by logusergithub** [comment]
> It is saved, but is not displayed when you next open the edit modal.

I would not have thought of that on my own. I see now the regex saved if I open the backup.json.

> This is true. Only OpenSearch engines get the shortcut option
I misunderstood at the beginning because when I made a copy the name is the same like when you do a shorcuts. You can imagine when I changed "{url}" inside 'Firefox' in the folder and used the second 'Firefox' not in the folder during the test. :-)

**Comment by karinaardos** [comment]
There is a small problem.
If there are two {searchTerms} in the argument, then only one is passed to the program as an argument

for example.
If the path is like this
d:\PotPlayerMini.exe n:\{searchTerms}\{searchTerms}_1080P.ts" /new

Transmitted like this
d:\PotPlayerMini.exe n:\selectText\{searchTerms}_1080P.ts" /new

Should be sent like this
d:\PotPlayerMini.exe n:\selectText\selectText_1080P.ts" /new

Can this be implemented?

**Comment by ssborbis** [comment]
> Can this be implemented?

I think it's just moving from .replace() to .replaceAll(). I never anticipated the terms being used more than once. Easy fix.

**Comment by ssborbis** [comment]
Here, this should do it
[contextsearch_web_ext-1.38a1-an+fx.xpi.zip](https://github.com/ssborbis/ContextSearch-web-ext/files/8194283/contextsearch_web_ext-1.38a1-an%2Bfx.xpi.zip)
.

**Comment by karinaardos** [comment]
Works great, thanks for the fast work

**Comment by ssborbis** [comment]
```
"C:\Program Files\DAUM\PotPlayer\PotPlayerMini64.exe" /add {searchTerms}
```

Works for me. With quotes, it would add a bunch of window .wav files for some reason.

> unlike the other dedicated extension I always used, that both the programs close themself during streaming after I close Firefox

I'll look into it. You might be able to get around it using `cmd /k` in your command line to open a new terminal

**Comment by logusergithub** [comment]
> Works for me. With quotes, it would add a bunch of window .wav files for some reason.

Thanks. I removed the quotes and the playlist was added, however seems it does't affect the closing behaviour of the program during streaming. As usual if you need my config file I'm available. I almost feel guilty about finding odd issues.

**Comment by ssborbis** [comment]
Actually, I'm still seeing the apps close. I'll need some time to see what's going on.

**Comment by logusergithub** [comment]
Thanks for the note. I was about to start over 2.10 again to confirm if I had something wrong and I was ready to make a video. I see that you are engaged in a work in progress on 2.11. I also noticed that PotPlayer /add works now with quotes on {searchTerms}.

I was also thinking about the {url} parameter: if it is actually useful when you open the external application from the toolbar, it places limitations if you use it from the context menu obviously outside 'page' search (links, audio, video) so you have to use a second App loader with {searchTerms}. To fully simulate dedicated extensions, which themselves are simpler than the useful complexity of CS, the {url} parameter should act as {SearchTerm} in the context menu if possible. But it's a little nuance not so important.

**Comment by ssborbis** [comment]
I think I have a fix for the closing when the browser closes. Please try thi
[ContextSearch.py.zip](https://github.com/ssborbis/ContextSearch-web-ext/files/8237299/ContextSearch.py.zip)

It's a windows issue. I never had that problem on linux.

**Comment by logusergithub** [comment]
Tested on multiple sites, it works. As always, thanks. 

From what I see Windows 10 will be the last and when the time comes I'll switch to Linux.

**Comment by ssborbis** [comment]
I am continuing to make changes to the ContextSearch.py file due to differences in how python behaves on linux and windows. If you want to update to the latest version, you can open the console from the CS options page and type this command:

```
checkAndUpdateNativeApp()
```

It will check the repo and prompt you to update if one is available. On the next version of CS, I'll add a button.

**Comment by logusergithub** [comment]
I noticed this as I was on CS finally finishing adding the latest search engines. 2.14 works fine as the last tested file.
I was just thinking about adding the NativeApp to my AutoHotkey script which I already use to update some filters on GitHub, but you anticipated me. Sure the future button is a useful new feature.

**Comment by justsome12** [comment]
this is so awesome. is it possible to open an image in paint? eg "https://upload.wikimedia.org/wikipedia/commons/thumb/3/3c/Wiki-lifebelt.svg/140px-Wiki-lifebelt.svg.png" with "C:/Windows/SysWOW64/mspaint.exe"

**Comment by ssborbis** [comment]
@justsome12 

Using the latest beta and native app:
`"C:/Windows/SysWOW64/mspaint.exe" "{download_url}"`

**Comment by rediffusion** [comment]
I use the Total Commander for opening everything I want. It replaces the main Windows 11 interface.
This whole thing is just messing around.

Sincerely!

**Comment by karinaardos** [comment]
Ability to select a folder to save the file.

in this form, the path is prescribed permanent
ffmpeg.exe -i "http://streamvideo/{searchTerms}" copy "e:\record\{searchTerms}.ts"

please add a new template {SELECT_PATH}
ffmpeg.exe -i "http://streamvideo/{searchTerms}" copy "e:\{SELECT_PATH}\{searchTerms}.ts"

To be able to select a folder to save the file.
Or is it impossible?

**Comment by ssborbis** [comment]
@karinaardos 

I'll have to see what the limitations are for the download API, but I think it can be done. I already have something in the works with the variable {ASK} which `prompt()`s for a string, not a file browser. That could work for your case, but a file browser would be better.

**Comment by karinaardos** [comment]
That would be totally cool. Thank you.

**Comment by ssborbis** [comment]
This thread is ancient, but the changes i mentioned in my last post are in the latest release ( 1.46.17 )

`{download_url=ask}` gets you the browser UI SaveAs dialog
`{prompt="what do you want to set this to?"}` brings up the browser `prompt()` dialog and allows you to type in a value for a parameter in your `path`

Example:

I have this template for opening an image in the DarkTable app, which asks me where I want to save the image
`darktable "{download_url=ask}"`

Also, I have this for downloading mp3s from youtube using yt-dlp, which asks what folder I want to download to
`gnome-terminal -- ~/bin/yt-dlp -P ~/Desktop/{prompt="Folder"} -x --audio-format mp3 "{searchTerms}"`

**Comment by karinaardos** [comment]
For some reason the Browser UI Saveas window does not appear. Instead, info windows appears with the contents of the command itself

**Comment by ssborbis** [comment]
Some test code got left in the release. New version going up now 1.46.18

> For some reason the Browser UI Saveas window does not appear

What is your template path?

**Comment by ssborbis** [comment]
Oh, and if you're using the Test button, the SaveAs (`{download_url=ask}') might not work. I'll have to look into why.

**Comment by karinaardos** [comment]
look at the screenshots
https://i.onthe.io/smngoz6gumgh6hjeh.f2a41b69.png
https://i.onthe.io/smngoz3t1fcko08bcg.5326df69.png

**Comment by ssborbis** [comment]
Ah, this might be due to not having the `downloads` permission. Try setting it manually until i get code that checks for it.

![image](https://github.com/ssborbis/ContextSearch-web-ext/assets/24310997/10157c53-47b5-45eb-8482-1fd7cea970d5)

**Comment by karinaardos** [comment]
I already gave all the permissions myself, but nothing has changed

**Comment by ssborbis** [comment]
Hmm. What browser? I've tested on FF, chromium and vivaldi

**Comment by karinaardos** [comment]
firefox 117.0.1

**Comment by ssborbis** [comment]
If you could some time, you might check for errors when inspecting the background page via about:debugging. 

Also, maybe double-check that clicking the "request permission" button on this page `options.html?permission=downloads#requestpermissions` does nothing ( because you already have it granted )

**Comment by karinaardos** [comment]
again I requested permissions through the address bar.
about:debugging - I don't understand what needs to be done.

Don’t waste your time on me, maybe I have something wrong with Firefox, although I have it almost new and I haven’t changed anything in my profile - strange

**Comment by ssborbis** [comment]
I did find a bug in the `{prompt=...}` code. I'll get an update out in a sec. I also added the download permissions check. Not sure why you're not seeing the SaveAs dialog. Which menu are you using to open that particular app launcher?

**Comment by karinaardos** [comment]
{prompt="what do you want to set this to?"} This worked for me before the update.
{download_url=ask} after the update, now nothing happens at all, the window with the command is no longer shown

I created a new firefox profile and installed only your add-on and it didn't help.
if you want, I will send you a zip archive of this profile, just give me the address where to send it

**Comment by ssborbis** [comment]
> I created a new firefox profile and installed only your add-on and it didn't help.

Yeah, I dunno what's going on there. I've had good results on all browsers I've tested. Are you on Windows, linux or macOS?

**Comment by karinaardos** [comment]
Windows

**Comment by ssborbis** [comment]
Ok, I'll see about loading a vbox and testing

**Comment by ssborbis** [comment]
It's working fine for me. Ok, try this for a sanity check

Make a new app launcher, use `{download_url=ask}` as the path
![image](https://github.com/ssborbis/ContextSearch-web-ext/assets/24310997/71c5edbd-6788-4549-b8c2-304a601a136c)

Save, and press the Test button. In the prompt, enter `http://example.com` and press ok
![image](https://github.com/ssborbis/ContextSearch-web-ext/assets/24310997/051cf539-ca0b-4bfa-b21e-16698fb059eb)

Do you get a SaveAs dialog box?

**Comment by karinaardos** [comment]
yes, now there is.

**Comment by ssborbis** [comment]
Ok, that's good. 

Newest version 1.46.20 going up now. It has some additional fixes for the {prompt...} feature.

Once updated, you might try a new engine with `{prompt="hello world"}` as the command and see what happens with the test button.

**Comment by ssborbis** [comment]
Ah, ok I found a bug in the SaveAs code. It would download the file but not open the app. I was relying on browser code that is only available in the developer edition. New code is in the git master branch, and the fix will be in the next release.

**Comment by ssborbis** [comment]
Fix is up in 1.46.22. I'll close out this thread, so if you have any further issues with SaveAs or the new {prompt...} params, open a new one.
